### PR TITLE
Verify request is not > possible PV size.

### PR DESCRIPTION
**Pull request by awels**
Verify that the PVC request size is not > possible PV size before creating PV.

Signed-off-by: Alexander Wels <awels@redhat.com>

### CODE REVIEW

File: hostpath-provisioner.go [hunk 1]
```diff
@@ -94,6 +94,15 @@ func isCorrectNode(annotations map[string]string, nodeName string) bool {
...
+	if pvCapacity != nil && pvCapacity.Cmp(pvc.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]) < 0 {
+		glog.Error("PVC request size larger than total possible PV size")
+		return false
+	}
+	return true;
+}
+
 // Provision creates a storage asset and returns a PV object representing it.
 func (p *hostPathProvisioner) Provision(options controller.ProvisionOptions) (*v1.PersistentVolume, error) {
 	isThisNode := isCorrectNode(options.PVC.GetAnnotations(), p.nodeName)
```
Comment (awels) [inline hostpath-provisioner.go]: Should be move this in there as well?

File: hostpath-provisioner.go [hunk 2]
```diff
@@ -94,12 +94,21 @@ func isCorrectNode(annotations map[string]string, nodeName string) bool {
 	return false
 }
 
+func (p *hostPathProvisioner) ShouldProvision(pvc *v1.PersistentVolumeClaim) bool {
```
Comment (j-griffith) [inline hostpath-provisioner.go]: I would've thought the qualifer would need declared for this.  If this works though that's great.

Comment (awels) [reply hostpath-provisioner.go]: I am not quite sure what you mean by qualifier, but yes this works.

File: hostpath-provisioner.go [hunk 3]
```diff
@@ -177,18 +189,19 @@ func (p *hostPathProvisioner) Delete(volume *v1.PersistentVolume) error {
 }
 
 func calculatePvCapacity(path string) (*resource.Quantity, error) {
-	var stat syscall.Statfs_t
-	err := syscall.Statfs(path, &stat)
+	statfs := &unix.Statfs_t{}
```
Comment (j-griffith) [inline hostpath-provisioner.go]: yeah, that's better on my side

**Comment by j-griffith** [comment]
This works better on my side; I'm not a fan of merging anything here without unit tests.  I'm working on adding the framework and the tests for the existing functions right now.  The question is would you be ok with waiting to merge this until that's up, you can rebase and add your tests?

**Comment by awels** [comment]
Absolutely, I am refining the calculation a little bit as well.

File: hostpath-provisioner.go [hunk 4]
```diff
@@ -94,12 +103,24 @@ func isCorrectNode(annotations map[string]string, nodeName string) bool {
...
 
+func (p *hostPathProvisioner) ShouldProvision(pvc *v1.PersistentVolumeClaim) bool {
+	shouldProvision := isCorrectNode(pvc.GetAnnotations(), p.nodeName)
+	
+	if shouldProvision {
+		pvCapacity, err := calculatePvCapacity(p.pvDir)
+		if pvCapacity != nil && pvCapacity.Cmp(pvc.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]) < 0 {
+			glog.Error("PVC request size larger than total possible PV size")
+			shouldProvision = false;
+		} else if pvCapacity == nil {
```
Comment (j-griffith) [inline hostpath-provisioner.go]: Why not short circuit this with `if err != nil`

File: hostpath-provisioner.go [hunk 5]
```diff
@@ -177,21 +198,40 @@ func (p *hostPathProvisioner) Delete(volume *v1.PersistentVolume) error {
...
 	}
-	totalSize := stat.Blocks * uint64(stat.Bsize)
-	if totalSize > math.MaxInt64 {
-		glog.Errorf("Calculated total disk size larger than: %d", math.MaxInt64)
-		return nil, errors.New("Total available space doesn't fit in int64")
+
+	// Capacity is total block count * block size
+	capacity := int64(statfs.Blocks) * int64(statfs.Bsize)
+
+	if err != nil {
```
Comment (j-griffith) [inline hostpath-provisioner.go]: There's no new assignment to err from L#202, so do you need this for anything?

File: hostpath-provisioner_test.go [hunk 6]
```diff
@@ -79,13 +77,33 @@ func Test_calculatePvCapacity(t *testing.T) {
 	type args struct {
 		path string
 	}
+	// Get total size of wherever we are running.
+	capacity, _ := getTotalCapacity(".")
```
Comment (j-griffith) [inline hostpath-provisioner_test.go]: This works, but aren't you kind of relying on the function your testing to determine the results of the test?

File: hostpath-provisioner_test.go [hunk 7]
```diff
@@ -94,9 +112,74 @@ func Test_calculatePvCapacity(t *testing.T) {
...
+
+func getTotalCapacity(path string) (int64, error) {
+	statfs := &unix.Statfs_t{}
+	err := unix.Statfs(path, statfs)
+	if err != nil {
+		return int64(-1), err
+	}
+
+	// Capacity is total block count * block size
+	return int64(statfs.Blocks) * int64(statfs.Bsize), nil
```
Comment (j-griffith) [inline hostpath-provisioner_test.go]: Bsize is already an int64; although it doesn't hurt to explicitly add it here.

File: hostpath-provisioner.go (hunk 5 above)
Comment (awels) [reply hostpath-provisioner.go]: Nope

File: hostpath-provisioner.go (hunk 4 above)
Comment (awels) [reply hostpath-provisioner.go]: Done

File: hostpath-provisioner.go (hunk 2 above)
Comment (j-griffith) [reply hostpath-provisioner.go]: I was wrong, don't worry about it :)  I was wondering if it needed a declaration or somethign specifically ton indicate it was part of the qualifier interface, but it looks like that all bubbles up through the volume package.

File: hostpath-provisioner_test.go [hunk 8]
```diff
@@ -94,9 +112,74 @@ func Test_calculatePvCapacity(t *testing.T) {
...
+func getTotalCapacity(path string) (int64, error) {
+	statfs := &unix.Statfs_t{}
+	err := unix.Statfs(path, statfs)
+	if err != nil {
+		return int64(-1), err
+	}
+
+	// Capacity is total block count * block size
+	return int64(statfs.Blocks) * statfs.Bsize, nil
+}
```
Comment (j-griffith) [inline hostpath-provisioner_test.go]: artifact?

Comment (awels) [reply hostpath-provisioner_test.go]: Meh its complaining there is no newline at the end..

**Pull request merged**